(feat: hook-store): create bundle hooks tenderly simulation

yvesfracari opened the pull request:
# Summary

Add bundle simulations on the hook store feature. This enhances the hooks simulations by considering all the previous hooks and the swap (in case of post hooks). 

https://www.loom.com/share/37cf070b4f7b41f896966c469941c406?sid=c183f6f3-f611-4d2f-8d7d-0015bb8d374a

# To Test

 Since we still don't have a Tenderly account, it is expected to have to login on Tenderly to see the simulation links.

1. Run the bff locally on port 3001. For more information check: https://github.com/cowprotocol/bff/pull/88 
2. Change the bff connection by editing the file: `libs/common-const/src/bff.ts`. Change the default URL to `http://localhost:3001`
3. Run the cowswap app locally.

4. Go to the hooks tab of the swap widget and add some hooks. 
- [ ] Should be able to see Tenderly links for failing transactions.
- [ ] Should auto refresh simulations on hook or token changes.
- [ ] Check the simulation for different tokens to sell and buy.
- [ ] Should be able to create order even if hooks are falling.

# Background

@alfetopito @shoom3301 In order to merge this PR we will have to add the .env keys. Let me know if you think that we should wait for the refactoring to the backend for the frontend repo.

alfetopito commented on apps/cowswap-frontend/src/modules/hooksStore/pure/AppliedHookItem/styled.tsx:
```diff
@@ -212,3 +212,19 @@ export const SimulateFooter = styled.div`
   }
 `
+
+export const Spinner = styled.div`
```
Could the existing spinner be used here?

https://github.com/cowprotocol/cowswap/blob/59cbc17225b825ee17ba0a9f93cfe3176fb2ca39/apps/explorer/src/components/common/Spinner.tsx#L1-L0

alfetopito replied:
Oh my 🤦 

Please disregard, I knew we had it but didn't notice it was from the Explorer

yvesfracari commented on apps/cowswap-frontend/package.json:
```diff
@@ -27,7 +27,9 @@
     ]
   },
-  "dependencies": {},
+  "dependencies": {
+    "@covalenthq/client-sdk": "^2.1.1"
+  },
```
Actually I will remove this pacakge, we don't need it

shoom3301 commented on apps/cowswap-frontend/src/modules/tenderly/const.ts:
```diff
@@ -0,0 +1,25 @@
+
+import { Chain, ChainName } from '@covalenthq/client-sdk'
+
+// Sorry Safe, you need to set up CORS policy :)
```
It was my comment :)

yvesfracari commented:
Hey everyone! I've refactored this to work with the `bff` and updated the PR description to include new instructions for testing it.

However, I am wondering if the bundle simulation feature would work better on a separate page after the swap definition. I believe we should revalidate on token amount changes, but this would trigger many requests in my opinion. For now, I'm just revalidating the simulation on token and hook changes (not amounts).

I look forward to your ideas.

yvesfracari commented on apps/cowswap-frontend/src/modules/tenderly/hooks/useTenderlyBundleSimulation.ts:
```diff
@@ -0,0 +1,69 @@
+  const tokenBuy = useTokenContract(orderParams?.buyTokenAddress)
+  const buyAmount = orderParams?.buyAmount
+
+  const { data: buyTokenTopHolders, isValidating: isTopTokenHoldersValidating } = useTopTokenHolders({
+    tokenAddress: tokenBuy?.address,
+    chainId,
+  })
```
We just use it when there are post hooks. However, I thought that running two requests in sequence could be a little slow for the user (find the holders and run the simulation). Also, the bff endpoint has some caches, so I think that this will not be so expensive.

alfetopito replied:
> We just use it when there are post hooks.

Even if this is used only when there are post hooks, it'll be called on every simulation even if not needed, correct?
And sure, it'll be cached and not hit upstream again unless the buy token changes, but it will still be called needlessly when there are no post hooks, no?
We could for example trigger it when the user opens the post-hook dialog, if the issue is the speed.

yvesfracari replied:
Makes sense! refactored the useSwr to a jotai atom with storage and timed cache. With that we will only do this request when we need it.

alfetopito commented on apps/cowswap-frontend/src/modules/tenderly/const.ts:
```diff
@@ -0,0 +1,6 @@
+const TENDERLY_ORG_NAME = process.env.REACT_APP_TENDERLY_ORG_NAME
+const TENDERLY_PROJECT_NAME = process.env.REACT_APP_TENDERLY_PROJECT_NAME
+
+export const getSimulationLink = (simulationId: string): string => {
+  return `https://dashboard.tenderly.co/${TENDERLY_ORG_NAME}/${TENDERLY_PROJECT_NAME}/simulator/${simulationId}`
+}
```
Are those still required? I thought we would get everything from BFF.
Like, return the whole simulation URL from the simulation endpoint.

alfetopito commented on apps/cowswap-frontend/src/modules/tenderly/hooks/useTenderlyBundleSimulation.ts:
```diff
@@ -0,0 +1,69 @@
+import { useHooks } from 'modules/hooksStore'
+import { generateNewSimulationData, generateSimulationDataToError } from '../utils/generateSimulationData'
+
+export function useTenderlyBundleSimulateSWR() {
```
One more thing I just noticed: file name doesn't match the exported function

alfetopito commented on apps/cowswap-frontend/src/modules/tenderly/hooks/useTenderlyBundleSimulation.ts:
```diff
@@ -0,0 +1,106 @@
+import { generateNewSimulationData, generateSimulationDataToError } from '../utils/generateSimulationData'
+import { getTokenTransferInfo } from '../utils/getTokenTransferInfo'
+
+export function useTenderlyBundleSimulate() {
```
The fn name still doesn't match 😬 

`useTenderlyBundleSimulation.ts` vs `useTenderlyBundleSimulate`

alfetopito reviewed (approved):
One more comment, otherwise looks good

shoom3301 commented on apps/cowswap-frontend/src/modules/tenderly/state/topTokenHolders.ts:
```diff
@@ -0,0 +1,55 @@
+    const currentTime = Date.now()
+
+    // 1 hour in milliseconds
+    if (cachedData[key] && currentTime - cachedData[key].timestamp <= 3600000) {
```
3600000 - please move to a const and use `ms` from `ms.macro`
